### PR TITLE
support winrt

### PR DESCRIPTION
**Pull request by john-peterson**
port a few things to winrt
## winrt detection

is probably correct but if i wanna verify it i wanna find winapifamily.h. this is an old winapifamily.h https://github.com/jp9000/OBS/blob/master/extras/winapifamily.h that still has WINAPI_FAMILY_ONE_PARTITION. i cant find the new one online

### CODE REVIEW

File: include/libtorrent/alert_types.hpp [hunk 1]
```diff
@@ -1338,6 +1338,7 @@ namespace libtorrent
 	// Port 0 asks the operating system to pick a port that's free). If that fails
 	// you may see a listen_failed_alert with port 0 even if you didn't ask to
 	// listen on it.
+#undef interface
```
Comment (arvidn) [inline include/libtorrent/alert_types.hpp]: could you leave a comment here explaining why this needs to be undefined? I take it some winrt header defines it. Is there no way to make it shut-up? similar to WIN32_MEAN_AND_LEAN.

Also, would it work if this moved into libtorrent/config.hpp?

**Comment by john-peterson** [comment]
> could you leave a comment here explaining why this needs to be undefined?

no need because it's common knowledge https://msdn.microsoft.com/en-us/library/737cydt1.aspx. it's blue in the vs editor like other keywords

> Is there no way to make it shut-up? similar to WIN32_MEAN_AND_LEAN.

probably not because it's a keyword like const, namespace etc

> Also, would it work if this moved into libtorrent/config.hpp?

no it cant be undef from another file. but i moved it to the top of the file so it looks a little better

**Comment by arvidn** [comment]
I would challenge the notion that it's common knowledge. I did no know about it for instance.

what happens if it's undefined in a different file? is it not a regular preprocessor define?

**Comment by john-peterson** [comment]
> I would challenge the notion that it's common knowledge. I did no know about it for instance.

we'll confuse people then with a mysterious undefine

> what happens if it's undefined in a different file?

if placed directly after "#include "libtorrent/config.hpp"" it still works. if placed at the bottom (directly before "#endif // TORRENT_CONFIG_HPP_INCLUDED") of it it doesn't

> is it not a regular preprocessor define?

cant cl define it for all files? it probably comes from /ZW

**Comment by arvidn** [comment]
the problem with not leaving a reason for it to be there, someone may come around later and remove it (and it would still work on almost all platforms).

I guess this header file gets included by clients, so it's not possible to control whether "interface" will be defined or not. A problem with just undefining it though is that it may break user code that's using it after including libtorrent/alert_types.hpp

I wonder if there's a way to prevent macro substitution there..

**Comment by john-peterson** [comment]
> the problem with not leaving a reason for it to be there, someone may come around later and remove it (and it would still works on almost all platforms).

k ill i changed the commit to explain it

> A problem with just undefining it though is that it may break user code that's using it after including libtorrent/alert_types.hpp

good point. i didnt think of that. **edit** however it might not propagate to files that include libtorrent/alert_types.hpp, since it's not propagated to libtorrent/alert_types.hpp from libtorrent/config.hpp

**Comment by arvidn** [comment]
thanks for looking into porting to winrt btw. It's something I've been tempted to give a stab for a while, but never gotten around to.

are these all the changes you need to make libtorrent build on winrt btw?

**Comment by john-peterson** [comment]
no there's tons of other stuff https://github.com/mirror/libtorrent/compare/master...john-peterson:rt

it might be why the app crashes instantly https://github.com/qbittorrent/qBittorrent/pull/2684#issuecomment-78177473

do u know how to "winrtrunner --debug" so i can get some output from the app to find why it crashes? i tried

```
winrtrunner --debug msvc src\release\qbittorrent.exe
```

but it doesn't say that msvc is bad debugger or give suggestions so i dunno what to write

**Comment by arvidn** [comment]
I've only done a little bit of winrt work (early days of porting uTorrent). Can't you just run it under the visual studio debugger?

**Comment by john-peterson** [comment]
there's no vs project for qtorrent

**Comment by arvidn** [comment]
you could try devenv.exe /debugexe qbittorrent.exe

I don't know if that works for winrt, but for normal programs it does

**Comment by john-peterson** [comment]
didnt do anything https://www.dropbox.com/s/cbfxra49lgi1a9l/Screenshot%202015-03-11%2006.08.20.png?dl=0

**Pull request closed without merging**